src/Shared/TryParseMethodCache.cs

```diff
@@ -87,32 +115,59 @@ public static bool HasTryParseMethod(ParameterInfo parameter)
                 return null;
             }
 
-            return MethodCallCache.GetOrAdd(type, Finder);
+            return _methodCallCache.GetOrAdd(type, Finder);
         }
 
-        private static MethodInfo GetEnumTryParseMethod()
+        private static MethodInfo GetEnumTryParseMethod(bool preferNonGenericEnumParseOverload)
         {
             var staticEnumMethods = typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static);
 
+            // With NativeAOT, if there's no static usage of Enum.TryParse<T>, it will be removed
+            // we fallback to the non-generic version if that is the case
```

Comment:
Is there a reason the non-generic one does not get trimmed?

Reply:
I assume it's because something else uses it. I have no idea what.

Reply:
Should we force it to be kept (e.g using `DynamicDependency`)

Reply:
I'm not sure if it's needed TBH, I *think* `typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static)` is visible to the linker but I'd need to make sure.

cc @MichalStrehovsky @eerhardt

Reply:
This is surprising to me. 

```C#
var staticEnumMethods = typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static);
```

That should "mark" all public static methods on `Enum`, and the AOT compiler shouldn't trim any of them.

Is it possible to build up the `GetMethod` query correctly so you only select the one method you are looking for?

Reply:
We still can't use the generic version of the other reasons mentioned in the PR.

Reply:
> typeof(Enum).GetMethods(BindingFlags.Public | BindingFlags.Static) is visible to the linker but I'd need to make sure.

Yes. It is a bug in NativeAOT: https://github.com/dotnet/runtimelab/issues/1402

Once this bug gets fixed, this will be broken again.

FWIW, I think fallbacks like this are wrong way to fix trimming and AOT friendliness issues.

Reply:
I didn't realize this is going to return `TryParse<T>` if that method got reflection-rooted through other means. The fact that it's not just rooted from `GetMethod` is kind of a bug but it only happens because the compiler figured there's not much it can do there - by default the AOT compiler will try to enable metadata for the method and generate a shared instantiation (over reference types), but the constraints on `TryParse<T>` rule out reference types. Right now, because it can't make a usable method body, it will not generate the metadata either. But we can fix it to create the metadata.

Yes, this is likely going to fail at `MakeGenericMethod` later.

Reply:
> Once this bug gets fixed, this will be broken again.

Thanks. I'll look into that.

> FWIW, I think fallbacks like this are wrong way to fix trimming and AOT friendliness issues.

It just needs to work well enough (or barely work) to enable people to try it out and give feedback. Seems like we need something a bit more creative here (like a check for IsDynamicCodeCompiled maybe).

Reply:
IsDynamicCodeSupported is what I did in the DataContractSerializer as a temporary workaround while we figure things out: https://github.com/dotnet/runtimelab/pull/692/files#diff-5c36d3b2cc664a8e51d53ac77390c98eca051ca63b13443233afc38ccee5e65fR60-R64

Reply:
Yep, this is what I'll do. FWIW I have a pretty good handle on most of the AOT friendly problems that will plague ASP.NET Core (like JSON 😄 ) and have prototypes for how to tackle most of them.